### PR TITLE
Give channel descriptions in `flutter channel`, use branch instead of upstream for channel name

### PR DESCRIPTION
**Pull request by Hixie**
## How we determine the channel name

Historically, we used the current branch's upstream to figure out the current channel name. I have no idea why. I traced it back to https://github.com/flutter/flutter/pull/446/files where @abarth implement this and I reviewed that PR and left no comment on it at the time.

I think this is confusing. You can be on a branch and it tells you that your channel is different. That seems weird.

This PR changes the logic to uses the current branch as the channel name.

## How we display channels

The main reason this PR exists is to add channel descriptions to the `flutter channel` list:

```
ianh@burmese:~/dev/flutter/packages/flutter_tools$ flutter channel
Flutter channels:
  master (tip of tree, for contributors)
  main (tip of tree, follows master channel)
  beta (updated monthly, recommended for experienced users)
  stable (updated quarterly, for new users and for production app releases)
* foo_bar

Currently not on an official channel.
ianh@burmese:~/dev/flutter/packages/flutter_tools$
```

## Other changes

I made a few other changes while I was at it:

* If you're not on an official channel, we used to imply `--show-all`, but now we don't, we just show the official channels plus yours. This avoids flooding the screen in the case the user is on a weird channel and just wants to know what channel they're on.
* I made the tool more consistent about how it handles unofficial branches. Now it's always `[user branch]`.
* I slightly adjusted how unknown versions are rendered so it's clearer the version is unknown rather than just having the word "Unknown" floating in the output without context.
* Simplified some of the code.
* Made some of the tests more strict (checking all output rather than just some aspects of it).
* Changed the MockFlutterVersion to implement the FlutterVersion API more strictly.
* I made sure we escape the output to `.metadata` to avoid potential injection bugs (previously we just inlined the version and channel name verbatim with no escaping, which is super sketchy).
* Tweaked the help text for the `downgrade` command to be clearer.
* Removed some misleading text in some error messages.
* Made the `.metadata` generator consistent with the template file.
* Removed some obsolete code to do with the `dev` branch.

## Reviewer notes

I'm worried that there are implications to some of these changes that I am not aware of, so please don't assume I know what I'm doing when reviewing this code. :-)

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/channel.dart [hunk 1]
```diff
@@ -74,8 +73,7 @@ class ChannelCommand extends FlutterCommand {
       throwToolExit('List channels failed: $result$details', exitCode: result);
     }
 
-    final List<String> officialChannels = kOfficialChannels.toList();
-    final List<bool> availableChannels = List<bool>.filled(officialChannels.length, false);
+    final Set<String> availableChannels = <String>{};
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/channel.dart:76]: are sets ordered?

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/commands/channel.dart:76]: nvm, I guess we're never printing these in order

File: packages/flutter_tools/lib/src/commands/downgrade.dart [hunk 2]
```diff
@@ -153,23 +153,23 @@ class DowngradeCommand extends FlutterCommand {
...
-        'internal error. Please ensure that git is installed on your computer and '
-        'retry again.\nError: $error.'
+        '$humanReadableVersion.\n'
+        'Error: $error'
       );
     }
     try {
       await processUtils.run(
+        // The `--` bit (because it's followed by nothing) means that we don't actually change
+        // anything in the working tree, which avoids the need to first go into detached HEAD mode.
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/downgrade.dart:163]: ahh, thanks for the comment

File: packages/flutter_tools/lib/src/version.dart [hunk 3]
```diff
@@ -40,12 +43,24 @@ enum Channel {
...
 const Set<String> kOfficialChannels = <String>{
-  globals.kDefaultFrameworkChannel,
+  'master',
   'main',
   'beta',
   'stable',
 };
 
+const Map<String, String> kChannelDescriptions = <String, String>{
+  'master': 'tip of tree, for contributors',
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/version.dart:53]: I've found several contributors confused by the term "tip of tree", or "tree" in general with regards to a repository. What about: "latest development branch, for contributors"?

File: packages/flutter_tools/lib/src/version.dart [hunk 4]
```diff
@@ -709,8 +717,8 @@ class GitTagVersion {
     String gitRef = 'HEAD'
   }) {
     if (fetchTags) {
-      final String channel = _runGit('git rev-parse --abbrev-ref HEAD', processUtils, workingDirectory);
-      if (channel == 'dev' || channel == 'beta' || channel == 'stable') {
+      final String channel = _runGit('git symbolic-ref --short HEAD', processUtils, workingDirectory);
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/version.dart:720]: for posterity, was `git rev-parse --abbrev-ref HEAD` giving us the wrong string?

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/version.dart:720]: Also, looks like this is a new git sub-command after an internal refactor. I wonder if all of our users will have it in their git? https://git-scm.com/docs/git-symbolic-ref#_notes

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/version.dart:720]: Well, it was at least around in 2011, so I'm guessing it should be fine https://stackoverflow.com/questions/4986000/whats-the-recommended-usage-of-a-git-symbolic-reference

Comment (Hixie) [reply packages/flutter_tools/lib/src/version.dart:720]: As far as I can tell what we did before was just wrong and this command has existed for ages.
But it's git.
So who knows.
I certainly cannot say with certainty that this is correct.

File: packages/flutter_tools/lib/src/version.dart (hunk 3 above)
Comment (Hixie) [reply packages/flutter_tools/lib/src/version.dart:53]: done

**Review by christopherfujino — APPROVED** [review]
LGTM

**Pull request merged**